kernel/src/syscall/sigaltstack.rs

```diff
@@ -1,131 +1,109 @@
 
-    let new_stack = {
-        let stack = ctx.user_space().read_val::<stack_t>(sig_stack_addr)?;
-        SigStack::try_from(stack)?
+    let stack = stack_t {
+        ss_sp: old_stack.base(),
+        ss_flags: flags.bits() as _,
+        ss_size: old_stack.size(),
     };
 
-    debug!("new_stack = {:?}", new_stack);
-
-    *ctx.thread_local.sig_stack().borrow_mut() = Some(new_stack);
-
-    Ok(())
+    ctx.user_space()
+        .write_val::<stack_t>(old_sig_stack_addr, &stack)
 }
 
-#[derive(Debug, Clone, Copy, Pod)]
-#[repr(C)]
-struct stack_t {
-    // Base address of stack
-    sp: Vaddr,
-    flags: i32,
-    // Number of bytes in stack
-    size: usize,
-}
+pub(super) fn set_new_stack(stack: stack_t, ctx: &Context, sp: usize) -> Result<()> {
+    let mut old_stack = ctx.thread_local.sig_stack().borrow_mut();
 
-impl TryFrom<stack_t> for SigStack {
-    type Error = Error;
-
-    fn try_from(stack: stack_t) -> Result<Self> {
-        if stack.flags < 0 {
-            return_errno_with_message!(Errno::EINVAL, "negative flags");
-        }
+    if old_stack.contains(sp) {
+        return_errno_with_message!(Errno::EPERM, "the old stack is active now");
+    }
 
-        let mut flags = SigStackFlags::from_bits(stack.flags as u32)
-            .ok_or_else(|| Error::with_message(Errno::EINVAL, "invalid flags"))?;
+    let flags = check_new_ss_flags(stack.ss_flags as u32)?;
 
-        if flags.contains(SigStackFlags::SS_DISABLE) {
-            return Ok(Self::new(0, flags, 0));
-        }
-        if stack.size < MINSTKSZ {
+    let new_stack = if flags.contains(SigStackFlags::SS_DISABLE) {
+        SigStack::new(0, flags, 0)
+    } else {
+        if stack.ss_size < MINSTKSZ {
             return_errno_with_message!(Errno::ENOMEM, "stack size is less than MINSTKSZ");
         }
-        if stack.sp.checked_add(stack.size).is_none() {
+
+        if stack.ss_sp.checked_add(stack.ss_size).is_none() {
             return_errno_with_message!(Errno::EINVAL, "overflow for given stack addr and size");
         }
 
-        if flags.is_empty() {
-            flags.insert(SigStackFlags::SS_ONSTACK);
-        }
+        SigStack::new(stack.ss_sp, flags, stack.ss_size)
+    };
 
-        Ok(Self::new(stack.sp, flags, stack.size))
-    }
+    debug!("new stack = {:x?}", new_stack);
+
+    *old_stack = new_stack;
+
+    Ok(())
 }
 
-impl From<SigStack> for stack_t {
-    fn from(stack: SigStack) -> Self {
-        let flags = stack.flags().bits() as i32 | stack.status() as i32;
+fn check_new_ss_flags(ss_flags: u32) -> Result<SigStackFlags> {
+    let ss_flags = SigStackFlags::from_bits(ss_flags)
+        .ok_or_else(|| Error::with_message(Errno::EINVAL, "unknown flags"))?;
 
-        Self {
-            sp: stack.base(),
-            flags,
-            size: stack.size(),
-        }
+    let status_flags = ss_flags & !SIG_STACK_FLAGS_MASK;
+    if status_flags != SigStackFlags::SS_DISABLE
+        && status_flags != SigStackFlags::SS_ONSTACK
```

Comment:
I don't believe we can set the `SS_ONSTACK` flag.

Reply:
Linux does allow the SS_ONSTACK flag to be set, although the reason why Linux permits this behavior is somewhat unclear to me.

https://elixir.bootlin.com/linux/v6.0.9/source/kernel/signal.c#L4193

Reply:
Looks like it is a BUG. The odd behavior is mentioned in the "BUGS" section of the man pages: https://man7.org/linux/man-pages/man2/sigaltstack.2.html#BUGS